### PR TITLE
SLING-6422, Allow for specifying oak restrictions with repoinit

### PR DESCRIPTION
**Pull request by nitin-nizhawan**
- Interpret parsed restriction clauses from repoinit parser

### CODE REVIEW

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 1]
```diff
@@ -54,8 +62,67 @@ public static JackrabbitAccessControlManager getJACM(Session s) throws Unsupport
...
+
+           for(RestrictionClause rc : list){
+               String restrictionName = rc.getName();
+               int type = jacl.getRestrictionType(restrictionName);
+               Value[] values = new Value[rc.getValues().size()];
+               for(int i=0;i<values.length;i++) {
+                   values[i] = vf.createValue(rc.getValues().get(i),type);
+               }
+
+               if(values.length > 1) {
```
Comment (anchela) [inline bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:107]: the length of the value array of a mv-restriction could potentially be zero... so i think checking for values > 1 is a bug and will lead to indexoutofboundsexception for the empty array. i would suggest to change it to values.length == 1

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 2]
```diff
@@ -54,8 +62,67 @@ public static JackrabbitAccessControlManager getJACM(Session s) throws Unsupport
...
+    }
+
+
     public static void setAcl(Session session, List<String> principals, List<String> paths, List<String> privileges, boolean isAllow)
             throws UnsupportedRepositoryOperationException, RepositoryException {
+        setAcl(session,principals,paths,privileges,isAllow,Arrays.asList(new RestrictionClause[]{}));
+    }
+
+    public static void setAcl(Session session, List<String> principals, List<String> paths, List<String> privileges, boolean isAllow, List<RestrictionClause> restrictionClauses)
+            throws UnsupportedRepositoryOperationException, RepositoryException {
```
Comment (anchela) [inline bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:125]: declaring _UnsupportedRepositoryOperationException_ is redundant here... in the static 'getJACM' above you removed it (unrelated change). so i would suggest to drop it here as well for consistency

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 3]
```diff
@@ -54,8 +62,67 @@ public static JackrabbitAccessControlManager getJACM(Session s) throws Unsupport
...
+               }
+           }
+        }
+        return new LocalRestrictions(restrictions,mvrestrictions);
+
+    }
+
+
     public static void setAcl(Session session, List<String> principals, List<String> paths, List<String> privileges, boolean isAllow)
             throws UnsupportedRepositoryOperationException, RepositoryException {
```
Comment (anchela) [inline bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:120]: declaring UnsupportedRepositoryOperationException is redundant here... in the static 'getJACM' above you removed it (unrelated change). so i would suggest to drop it here as well for consistency

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 4]
```diff
@@ -64,8 +131,15 @@ public static void setAcl(Session session, List<String> principals, List<String>
             if(!session.nodeExists(path)) {
                 throw new PathNotFoundException("Cannot set ACL on non-existent path " + path);
             }
+
             JackrabbitAccessControlList acl = AccessControlUtils.getAccessControlList(session, path);
+
+            validateRestrictionClauses(restrictionClauses, acl);
+
+            LocalRestrictions oakRestrictions = createOakRestrictions(restrictionClauses,acl,session);
```
Comment (anchela) [inline bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:139]: why are they called 'oakRestrictions'? that looks a bit confusing to me as the restrictions are part of the jackrabbit api... i would drop the mentioning of oak here unless you are not making use of Oak API/SPI.

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 5]
```diff
@@ -81,12 +155,13 @@ public static void setAcl(Session session, List<String> principals, List<String>
                 if (principal == null) {
                     throw new IllegalStateException("Principal not found: " + name);
                 }
-                LocalAccessControlEntry newAce = new LocalAccessControlEntry(principal, jcrPriv, isAllow);
+                LocalAccessControlEntry newAce = new LocalAccessControlEntry(principal, jcrPriv, isAllow, oakRestrictions);
```
Comment (anchela) [inline bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:158]: same here...  i don't think that referring to oak for the restrictions is really correct here, is it?

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 6]
```diff
@@ -81,12 +155,13 @@ public static void setAcl(Session session, List<String> principals, List<String>
...
                 if (contains(existingAces, newAce)) {
                     LOG.info("Not adding {} to path {} since an equivalent access control entry already exists", newAce, path);
                     continue;
                 }
-                acl.addEntry(newAce.principal, newAce.privileges, newAce.isAllow);
+                acl.addEntry(newAce.principal, newAce.privileges, newAce.isAllow,
+                             newAce.oakRestrictions.getRestrictions(),newAce.oakRestrictions.getMVRestrictions());
```
Comment (anchela) [inline bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:164]: same here

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 7]
```diff
@@ -122,21 +197,60 @@ private static String toString(JackrabbitAccessControlEntry entry) throws Reposi
...
         private final boolean isAllow;
-        
+        private final LocalRestrictions oakRestrictions;
+
         LocalAccessControlEntry(Principal principal, Privilege[] privileges, boolean isAllow) {
+            this(principal, privileges, isAllow, null);
+        }
+
+        LocalAccessControlEntry(Principal principal, Privilege[] privileges,
+                                boolean isAllow, LocalRestrictions oakRestrictions) {
```
Comment (anchela) [inline bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:207]: and here.
btw: what does 'local' stands for? and local to what?

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 8]
```diff
@@ -122,21 +197,60 @@ private static String toString(JackrabbitAccessControlEntry entry) throws Reposi
...
             this.principal = principal;
             this.privileges = privileges;
             this.isAllow = isAllow;
+            this.oakRestrictions = oakRestrictions != null ? oakRestrictions : new LocalRestrictions();
```
Comment (anchela) [inline bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:211]: same again

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 9]
```diff
@@ -122,21 +197,60 @@ private static String toString(JackrabbitAccessControlEntry entry) throws Reposi
...
+                        }
+                    }
+                }
+                return true;
+            }
+            return false;
+        }
+
         private boolean contains(Privilege[] first, Privilege[] second) {
             // we need to ensure that the privilege order is not taken into account, so we use sets
```
Comment (anchela) [inline bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:255]: that's not your fault... but this method has a major bug as it doesn't take into account that there are aggregated privileges. a set that just compares privileges won't work.... this method needs to take the aggregation into account if it wants to be correct.

a simple test would have revealed this: take privilege jcr:read and test if rep:readNodes is contained. or alternatively: jcr:modifyProperities or jcr:write (which is even defined an aggregated in the JCR specification) or jcr:all (the most obvious).

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 10]
```diff
@@ -54,8 +62,67 @@ public static JackrabbitAccessControlManager getJACM(Session s) throws Unsupport
...
 
+    /**
+     * checks if provided restriction is supported at
+     * current path or not
+     *
+     * @param specifiedRestrictions
+     * @param jacl
+     * @return
+     */
+    private static void validateRestrictionClauses(List<RestrictionClause> specifiedRestrictions, JackrabbitAccessControlList jacl) throws RepositoryException{
```
Comment (anchela) [inline bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:73]: not sure if that is really needed... IMHO the implementation of the Oak RestrictionProvider is in charge of validating the restrictions latest upon save. i would rely on that instead of duplicating the validation, which IMO goes beyond just checking for supported names. in other words: a given restriction definition may still fail upon save if some other constraint is violated (e.g. passing wrong type or passing mv when single value is expected or if a mandatory restriction is missing).

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 11]
```diff
@@ -54,8 +62,67 @@ public static JackrabbitAccessControlManager getJACM(Session s) throws Unsupport
...
+     * Converts RestrictionClauses to structure consumable by
+     * jackrabbit
+     * @param list
+     * @param jacl
+     * @param s
+     * @return
+     * @throws RepositoryException
+     */
+    private static LocalRestrictions createOakRestrictions(List<RestrictionClause> list, JackrabbitAccessControlList jacl, Session s) throws RepositoryException {
+        Map<String,Value> restrictions = new HashMap<String,Value>();
```
Comment (anchela) [inline bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:93]: i would drop 'oak' from the method name. first of all you are not creating Oak restrictions but only relying on Jackrabbit API and second you are effectively returning an "LocalRestriction", which in turn has no reference to Oak restriction API... so no reason to mention oak here. I would just call it "createRestrictions" or "createLocalRestrictions", which is what you are really doing, right?

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 12]
```diff
@@ -122,21 +197,60 @@ private static String toString(JackrabbitAccessControlEntry entry) throws Reposi
         private final Principal principal;
         private final Privilege[] privileges;
         private final boolean isAllow;
-        
+        private final LocalRestrictions oakRestrictions;
```
Comment (anchela) [inline bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:200]: and here again :-) these are not oak restrictions

File: bundles/jcr/repoinit/src/test/java/org/apache/sling/jcr/repoinit/GeneralAclTest.java [hunk 13]
```diff
@@ -400,4 +440,132 @@ public void mergePreserveMode_PreserveAceTest() throws Exception {
...
+        }
+    }
+
+
+    /**
+     * Tests use of rep:itemNames restriction in set ACL
+     * @throws Exception
+     */
+    @Test
+    public void itemNamesRestrictionTest() throws Exception {
```
Comment (anchela) [inline bundles/jcr/repoinit/src/test/java/org/apache/sling/jcr/repoinit/GeneralAclTest.java:485]: i would suggest to also create a test with empty-mv- restriction (see above for the minor bug wrt this)

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 1 above)
Comment (nitin-nizhawan) [reply bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:107]: Hi, @anchela I agree with you that this is a bug.  Issue here is that we do not know at this point whether the restriction is mv or not. So, we are creating mv restriction if there are more than 1 values.  Would you know, if jackrabbit has an API to find if a restriction is mv?

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 11 above)
Comment (nitin-nizhawan) [reply bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:93]: Ok, makes sense

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 10 above)
Comment (nitin-nizhawan) [reply bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:73]: Ok, I will remove it in that case

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 3 above)
Comment (nitin-nizhawan) [reply bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:120]: I think it got added after merging upstream changes, Anyway I will remove it

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 4 above)
Comment (nitin-nizhawan) [reply bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:139]: sure

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 2 above)
Comment (nitin-nizhawan) [reply bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:125]: will remove it

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 5 above)
Comment (nitin-nizhawan) [reply bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:158]: will remove it

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 12 above)
Comment (nitin-nizhawan) [reply bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:200]: Will rename it to "localRestrictions"

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 7 above)
Comment (nitin-nizhawan) [reply bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:207]: Local means local to this class (inspired byLocalAccessControEntry :)  ) , just to separate from "Restriction"

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 9 above)
Comment (nitin-nizhawan) [reply bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:255]: 👍  I will log a bug for this since it is unrelated to this change. Thanks for the test case

File: bundles/jcr/repoinit/src/test/java/org/apache/sling/jcr/repoinit/GeneralAclTest.java (hunk 13 above)
Comment (nitin-nizhawan) [reply bundles/jcr/repoinit/src/test/java/org/apache/sling/jcr/repoinit/GeneralAclTest.java:485]: ok

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 1 above)
Comment (anchela) [reply bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:107]: @nitin-nizhawan unfortunately that's still missing in the jackrabbit API... but my point wasn't primarily about the case when you have 1 value. it's about an empty value array. as long as there is no way to find out about the mv-status, i would test for value length being 1 (and relying on RestrictionProvider impls to deal with a single value if the target is multivalued).

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 9 above)
Comment (nitin-nizhawan) [reply bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:255]: Logged JIRA for this https://issues.apache.org/jira/browse/SLING-6867

File: bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 1 above)
Comment (nitin-nizhawan) [reply bundles/jcr/repoinit/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:107]: Oh thanks, @anchela I completely missed ioobe. I will update this.

**Comment by nitin-nizhawan** [comment]
Closing this pull request, will raise fresh pr
